File: auth-center/auth-center-domain/src/test/java/com/jmsoftware/maf/authcenter/permission/service/impl/PermissionDomainServiceImplTest.java
package com.jmsoftware.maf.authcenter.permission.service.impl;

import com.google.common.collect.Lists;
import com.jmsoftware.maf.authcenter.permission.mapper.PermissionMapper;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.parallel.Execution;
import org.junit.jupiter.api.parallel.ExecutionMode;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * <h1>PermissionDomainServiceImplTest</h1>
 * Description: PermissionDomainServiceImplTest, change description here.
 * <p>
 * <h2>Mockito JUnit 5 Extension</h2>
 * <p>There is also a Mockito extension for JUnit 5 that will make the initialization even simpler.</p>
 * <p><strong>Pros:</strong></p>
 * <ul>
 * <li>No need to call <code>MockitoAnnotations.openMocks()</code></li>
 * <li>Validates framework usage and detects incorrect stubbing</li>
 * <li>Easy to create mocks</li>
 * <li>Very readable</li>
 *
 * </ul>
 * <p><strong>Cons:</strong></p>
 * <ul>
 * <li>Need an extra dependency on <code>org.mockito:mockito-junit-jupiter</code>, which has been included by Spring.
 * So we don&#39;t have to worry about this.</li>
 * </ul>
 *
 * @author Johnny Miller (鍾俊), e-mail: johnnysviva@outlook.com, date: 4/4/2022 8:40 AM
 * @see <a href='https://www.arhohuttunen.com/junit-5-mockito/'>Using Mockito With JUnit 5</a>
 * @see <a href='https://www.youtube.com/watch?v=p7_cTAF39A8/'>YouTube - Using Mockito With JUnit 5</a>
 **/
@Slf4j
@ExtendWith(MockitoExtension.class)
@Execution(ExecutionMode.CONCURRENT)
class PermissionDomainServiceImplTest {
    @InjectMocks
    private PermissionDomainServiceImpl permissionDomainService;
    @Mock
    private PermissionMapper permissionMapper;

    @BeforeEach
    void setUp() {
        log.info("{} setUp", this.getClass().getSimpleName());
    }

    @AfterEach
    void tearDown() {
        log.info("{} tearDown", this.getClass().getSimpleName());
    }

    @Test
    void getPermissionListByRoleIdList() {
        when(this.permissionMapper.selectPermissionListByRoleIdList(anyList(), anyList()))
                .thenReturn(Lists.newArrayList());
        val permissionList = this.permissionDomainService.getPermissionListByRoleIdList(
                Lists.newArrayList(), Lists.newArrayList());
        log.info("permissionList: {}", permissionList);
        verify(this.permissionMapper).selectPermissionListByRoleIdList(anyList(), anyList());
        assertNotNull(permissionList);
    }
}

File: auth-center/auth-center-domain/src/test/java/com/jmsoftware/maf/authcenter/permission/service/impl/PermissionDomainServiceImplTest.kt
package com.jmsoftware.maf.authcenter.permission.service.impl

import com.jmsoftware.maf.authcenter.permission.mapper.PermissionMapper
import com.jmsoftware.maf.authcenter.permission.persistence.Permission
import com.jmsoftware.maf.common.util.logger
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.parallel.Execution
import org.junit.jupiter.api.parallel.ExecutionMode
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension

/**
 * # PermissionDomainServiceImplTest
 *
 * Description: PermissionDomainServiceImplTest, change description here.
 *
 * ## Mockito JUnit 5 Extension
 *
 * There is also a Mockito extension for JUnit 5 that will make the initialization even simpler.
 *
 * **Pros:**
 *
 *  * No need to call `MockitoAnnotations.openMocks()`
 *  * Validates framework usage and detects incorrect stubbing
 *  * Easy to create mocks
 *  * Very readable
 *
 * **Cons:**
 *
 *  * Need an extra dependency on `org.mockito:mockito-junit-jupiter`, which has been included by Spring.
 * So we don&#39;t have to worry about this.
 *
 *
 * @author Johnny Miller (鍾俊), e-mail: johnnysviva@outlook.com, date: 4/4/2022 8:40 AM
 * @see <a href='https://www.arhohuttunen.com/junit-5-mockito/'>Using Mockito with JUnit 5</a>
 * @see <a href='https://www.youtube.com/watch?v=p7_cTAF39A8/'>YouTube - Using Mockito with JUnit 5</a>
 */
@ExtendWith(MockitoExtension::class)
@Execution(ExecutionMode.CONCURRENT)
internal class PermissionDomainServiceImplTest {
    companion object {
        private val log = logger()
    }

    @InjectMocks
    private lateinit var permissionDomainService: PermissionDomainServiceImpl

    @Mock
    private lateinit var permissionMapper: PermissionMapper

    @BeforeEach
    fun setUp() {
        log.info("{} setUp", this.javaClass.simpleName)
    }

    @AfterEach
    fun tearDown() {
        log.info("{} tearDown", this.javaClass.simpleName)
    }

    @Test
    fun getPermissionListByRoleIdList() {
        val permission = Permission()
        permission.id = 1
        `when`(permissionMapper.selectPermissionListByRoleIdList(anyList(), anyList()))
            .thenReturn(listOf(permission))
        val permissionList = permissionDomainService.getPermissionListByRoleIdList(listOf(), listOf())
        log.info("permissionList: $permissionList")
        verify(permissionMapper)
            .selectPermissionListByRoleIdList(anyList(), anyList())
        assertNotNull(permissionList)
    }
}
